javaslang-benchmark/src/test/java/javaslang/benchmark/collection/TreeMapBenchmark.java

```diff
@@ -0,0 +1,398 @@
+public class TreeMapBenchmark {
+    private static final Ordering<Integer> SCALA_ORDERING = Ordering$.MODULE$.comparatorToOrdering(Integer::compareTo);
+
+    @Test
+    public void launchBenchmark() throws Exception {
+        Options opt = new OptionsBuilder()
+                .include(this.getClass().getSimpleName())
+                .shouldFailOnError(true)
+                .shouldDoGC(true)
+                .build();
+
+        Collection<RunResult> results = new Runner(opt).run();
+        BenchmarkPerformanceReporter reporter = BenchmarkPerformanceReporter.of(results);
+        reporter.printDetailedPerformanceReport();
+        reporter.printPerformanceRatiosReport();
+    }
+
+    // The JMH samples are the best documentation for how to use it
+    // http://hg.openjdk.java.net/code-tools/jmh/file/tip/jmh-samples/src/main/java/org/openjdk/jmh/samples/
+    @State(Scope.Thread)
+    public static class BenchmarkState {
+        @Param({"10", "1000", "100000"})
+        int numItems;
+
+        Random random = new Random();
+
+        int minKey;
+        int maxKey;
+
+        Array<Integer> keysSorted;
+        Array<Integer> keysRandom;
+        Array<Tuple2<Integer, String>> entriesRandom;
+
+        TreeMap<Integer, String> slangTreeMap = TreeMap.empty();
+        java.util.TreeMap<Integer, String> javaTreeMap = new java.util.TreeMap<>();
+        scala.collection.immutable.TreeMap<Integer, String> scalaTreeMap = new scala.collection.immutable.TreeMap<>(SCALA_ORDERING);
+
+        @Setup(Level.Iteration)
+        public void setUp() {
+            minKey = 1;
+            maxKey = numItems * 2;
+
+            // We exclude keys with odd numbers to test methods requiring missing keys, such as floor or ceiling
+            keysSorted = Array.rangeBy(minKey, maxKey, 2);
+            keysRandom = randomize(keysSorted);
+            entriesRandom = keysRandom.map(i -> Tuple.of(i, Integer.toString(i)));
+
+            entriesRandom.forEach(e -> {
+                slangTreeMap = slangTreeMap.put(e._1, e._2);
+                javaTreeMap.put(e._1, e._2);
+                scalaTreeMap = scalaTreeMap.$plus(new scala.Tuple2<>(e._1, e._2));
+            });
+        }
+
+        @TearDown
+        public void tearDown() {
+            keysRandom = null;
+            keysRandom = null;
+            entriesRandom = null;
+            slangTreeMap = null;
+            javaTreeMap = null;
+            scalaTreeMap = null;
+        }
+
+        // Valid keys are only even numbers
+        int nextValidRandomKey() {
+            return keysRandom.get(random.nextInt(numItems));
+        }
+
+        int nextRandomKey() {
+            return random.nextInt(maxKey);
+        }
+
+        Array<Integer> randomize(Array<Integer> list) {
+            java.util.List<Integer> integers = list.toJavaList();
+            Collections.shuffle(integers);
+            return Array.ofAll(integers);
+        }
+    }
+
+    @Benchmark
+    @Group("GetAndPut")
```

Comment:
please consider extracting groups to constants.
How do `Group`s work exactly, aren't they wormed up together? Are we sure we want that?
We can do the grouping in separate subclasses also, as done in `PriorityQueueBenchmark`.